cores/esp8266/core_esp8266_waveform.cpp

```diff
@@ -102,67 +95,52 @@ void setTimer1Callback(uint32_t (*fn)()) {
   timer1CB = fn;
   if (!timerRunning && fn) {
     initTimer();
-    timer1_write(microsecondsToClockCycles(1)); // Cause an interrupt post-haste
-  } else if (timerRunning && !fn && !waveformEnabled) {
+    timer1_write(microsecondsToClockCycles(2)); // Cause an interrupt post-haste
+  } else if (timerRunning && !fn && !waveformsEnabled) {
     deinitTimer();
   }
 }
 
+int startWaveform(uint8_t pin, uint32_t timeHighUS, uint32_t timeLowUS, uint32_t runTimeUS) {
+  return startWaveformCycles(pin,
+    microsecondsToClockCycles(timeHighUS), microsecondsToClockCycles(timeLowUS), microsecondsToClockCycles(runTimeUS));
+}
+
 // Start up a waveform on a pin, or change the current one.  Will change to the new
 // waveform smoothly on next low->high transition.  For immediate change, stopWaveform()
 // first, then it will immediately begin.
-int startWaveform(uint8_t pin, uint32_t timeHighUS, uint32_t timeLowUS, uint32_t runTimeUS) {
-  if ((pin > 16) || isFlashInterfacePin(pin)) {
+int startWaveformCycles(uint8_t pin, uint32_t timeHighCycles, uint32_t timeLowCycles, uint32_t runTimeCycles) {
+  if ((pin > 16) || isFlashInterfacePin(pin) || !(timeHighCycles + timeLowCycles)) {
     return false;
   }
-  Waveform *wave = &waveform[pin];
+  Waveform* wave = &waveforms[pin];
   // Adjust to shave off some of the IRQ time, approximately
-  wave->nextTimeHighCycles = microsecondsToClockCycles(timeHighUS);
-  wave->nextTimeLowCycles = microsecondsToClockCycles(timeLowUS);
-  wave->expiryCycle = runTimeUS ? GetCycleCount() + microsecondsToClockCycles(runTimeUS) : 0;
-  if (runTimeUS && !wave->expiryCycle) {
-    wave->expiryCycle = 1; // expiryCycle==0 means no timeout, so avoid setting it
-  }
+  wave->nextTimeHighCycles = timeHighCycles;
+  wave->nextTimeLowCycles = timeLowCycles;
 
-  uint32_t mask = 1<<pin;
-  if (!(waveformEnabled & mask)) {
+  uint32_t mask = 1UL << pin;
+  if (!(waveformsEnabled & mask)) {
     // Actually set the pin high or low in the IRQ service to guarantee times
-    wave->nextServiceCycle = GetCycleCount() + microsecondsToClockCycles(1);
-    waveformToEnable |= mask;
+    uint32_t now = ESP.getCycleCount();
```

Comment:
Could you check the assembly and make sure it's not actually doing a call here and just inlining the rdsr?  I think originally that was the reason it was explicitly added to the code here.

Reply:
umm. What? :-) I did read the code and decided that the embedded code was the same in effect as calling the APIs.